File: prototype-rust/progred_graph/src/mutgid.rs
use crate::gid::Gid;
use crate::id::Id;
use im::HashMap;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use uuid::Uuid;

#[derive(Debug, Clone)]
pub struct MutGid {
    data: HashMap<Uuid, HashMap<Id, Id>>,
}

impl MutGid {
    pub fn new() -> Self {
        Self {
            data: HashMap::new(),
        }
    }

    pub fn entities(&self) -> impl Iterator<Item = &Uuid> {
        self.data.keys()
    }

    pub fn ptr_eq(&self, other: &Self) -> bool {
        self.data.ptr_eq(&other.data)
    }
}

impl Gid for MutGid {
    fn edges(&self, entity: &Id) -> Option<&HashMap<Id, Id>> {
        match entity {
            Id::Uuid(uuid) => self.data.get(uuid),
            _ => None,
        }
    }
}

impl MutGid {
    pub fn set(&mut self, entity: Uuid, label: Id, value: Id) {
        let edges = match self.data.get(&entity) {
            Some(e) => e.update(label, value),
            None => HashMap::unit(label, value),
        };
        self.data.insert(entity, edges);
    }

    pub fn merge(&mut self, other: HashMap<Uuid, HashMap<Id, Id>>) {
        for (entity, new_edges) in other {
            let merged = match self.data.get(&entity) {
                Some(existing) => existing.clone().union(new_edges),
                None => new_edges,
            };
            self.data.insert(entity, merged);
        }
    }

    pub fn delete(&mut self, entity: &Uuid, label: &Id) {
        if let Some(edges) = self.data.get(entity) {
            let new_edges = edges.without(label);
            self.data = if new_edges.is_empty() {
                self.data.without(entity)
            } else {
                self.data.update(*entity, new_edges)
            };
        }
    }

    pub fn retain_entities(&mut self, keep: &std::collections::HashSet<Id>) {
        self.data = self.data.iter()
            .filter(|(uuid, _)| keep.contains(&Id::Uuid(**uuid)))
            .map(|(&k, v)| (k, v.clone()))
            .collect();
    }

    pub fn purge(&mut self, id: &Id) {
        if let Id::Uuid(uuid) = id {
            self.data = self.data.without(uuid);
        }
        self.data = self.data.iter()
            .map(|(&entity, edges)| {
                let filtered = edges.iter()
                    .filter(|(_, v)| *v != id)
                    .map(|(k, v)| (k.clone(), v.clone()))
                    .collect();
                (entity, filtered)
            })
            .filter(|(_, edges): &(_, im::HashMap<Id, Id>)| !edges.is_empty())
            .collect();
    }
}

impl Default for MutGid {
    fn default() -> Self {
        Self::new()
    }
}

impl Serialize for MutGid {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let proxy: std::collections::BTreeMap<Uuid, Vec<(Id, Id)>> = self.data
            .iter()
            .map(|(uuid, edges)| {
                let mut pairs: Vec<_> = edges.iter().map(|(k, v)| (k.clone(), v.clone())).collect();
                pairs.sort();
                (*uuid, pairs)
            })
            .collect();
        proxy.serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for MutGid {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let proxy: std::collections::HashMap<Uuid, Vec<(Id, Id)>> =
            std::collections::HashMap::deserialize(deserializer)?;
        let data = proxy.into_iter()
            .map(|(uuid, edges)| (uuid, edges.into_iter().collect()))
            .collect();
        Ok(MutGid { data })
    }
}

File: progred/progred/Graph/MutGid.swift
import Foundation
import HashTreeCollections

struct MutGid: Gid {
    private(set) var data: TreeDictionary<UUID, TreeDictionary<Id, Id>>

    init() {
        data = TreeDictionary()
    }

    var entities: some Sequence<UUID> {
        data.keys
    }

    func edges(entity: Id) -> Edges? {
        guard case .uuid(let uuid) = entity else { return nil }
        return data[uuid].map { Edges(data: $0, readOnly: false) }
    }

    mutating func set(entity: UUID, label: Id, value: Id) {
        var edges = data[entity] ?? TreeDictionary()
        edges[label] = value
        data[entity] = edges
    }

    mutating func commit(entity: UUID, label: Id, value: Id?) {
        if let value {
            set(entity: entity, label: label, value: value)
        } else {
            delete(entity: entity, label: label)
        }
    }

    mutating func delete(entity: UUID, label: Id) {
        guard var edges = data[entity] else { return }
        edges.removeValue(forKey: label)
        if edges.isEmpty {
            data.removeValue(forKey: entity)
        } else {
            data[entity] = edges
        }
    }

    mutating func merge(_ other: TreeDictionary<UUID, TreeDictionary<Id, Id>>) {
        for (entity, newEdges) in other {
            if var existing = data[entity] {
                existing.merge(newEdges, uniquingKeysWith: { _, new in new })
                data[entity] = existing
            } else {
                data[entity] = newEdges
            }
        }
    }

    func frozen() -> ImmGid {
        ImmGid(data: data)
    }

    mutating func retainEntities(_ keep: Set<Id>) {
        data = TreeDictionary(
            uniqueKeysWithValues: data.filter { key, _ in keep.contains(.uuid(key)) }
        )
    }

    mutating func purge(_ id: Id) {
        if case .uuid(let uuid) = id {
            data.removeValue(forKey: uuid)
        }
        data = TreeDictionary(
            uniqueKeysWithValues: data.compactMap { entity, edges -> (UUID, TreeDictionary<Id, Id>)? in
                let filtered = TreeDictionary(
                    uniqueKeysWithValues: edges.filter { _, v in v != id }
                )
                return filtered.isEmpty ? nil : (entity, filtered)
            }
        )
    }
}
